File: MovieApplication/app/src/main/java/com/ryan/movieapplication/RecyclerView/RecyclerViewAdapter.java
package com.ryan.movieapplication.RecyclerView;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.constraint.ConstraintLayout;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.ryan.movieapplication.R;
import com.ryan.movieapplication.RetroFit.MovieInfo;
import com.ryan.movieapplication.UserInterface.MovieItem;

import java.util.List;

public class RecyclerViewAdapter extends RecyclerView.Adapter<RecyclerViewAdapter.ViewHolder>
{
    private List<MovieInfo> movies;
    private Context context;

    public RecyclerViewAdapter(List<MovieInfo> movies, Context context)
    {
        this.movies = movies;
        this.context = context;
    }

    public class ViewHolder extends RecyclerView.ViewHolder
    {
        public ImageView imageView;
        public ConstraintLayout constraintLayout;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            imageView = itemView.findViewById(R.id.image_view);
            constraintLayout = itemView.findViewById(R.id.constraint_layout);
        }
    }
    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.recycler_view_image, viewGroup, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder viewHolder, final int i) {

        Uri uri = Uri.parse("https://image.tmdb.org/t/p/w185" + movies.get(i).getPosterPath());
        if(movies.get(i).getPosterPath() == null || movies.get(i).getPosterPath().isEmpty())
        {
            Glide.with(context).load(R.drawable.nomovieposter).into(viewHolder.imageView);
        }
        else
        {
            Glide.with(context).load(uri).into(viewHolder.imageView);
        }
        viewHolder.constraintLayout.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(context, MovieItem.class);
                intent.putExtra("title", movies.get(i).getTitle());
                intent.putExtra("overview", movies.get(i).getOverview());
                intent.putExtra("image", movies.get(i).getPosterPath());
                context.startActivity(intent);
            }
        });

    }

    @Override
    public int getItemCount() {
        return movies.size();
    }

    public void updateMovieInfoList(List<MovieInfo> newMovieInfoList)
    {
        final MoviesDiffCallback diffCallback = new MoviesDiffCallback(movies, newMovieInfoList);
        final DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(diffCallback);

        movies.clear();
        movies.addAll(newMovieInfoList);
        diffResult.dispatchUpdatesTo(this);
    }

}

File: RedditBrowser/app/src/main/java/com/ryan/redditbrowser/UI/Recyclerview/RecyclerViewAdapter.kt
package com.ryan.redditbrowser.UI.Recyclerview

import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.browser.customtabs.CustomTabsIntent
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.ryan.redditbrowser.R
import com.ryan.redditbrowser.Retrofit.Data.Children
import com.ryan.redditbrowser.UI.CustomTabHelper
import com.ryan.redditbrowser.UI.WebView
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.recyclerview_item.view.*


class RecyclerViewAdapter(var children: MutableList<Children>, val context: Context) :
    RecyclerView.Adapter<RecyclerViewAdapter.ViewHolder>() {

    private var customTabHelper: CustomTabHelper = CustomTabHelper()

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): RecyclerViewAdapter.ViewHolder {
        return ViewHolder(
            LayoutInflater.from(context).inflate(
                R.layout.recyclerview_item,
                parent,
                false
            )
        )
    }

    override fun getItemCount(): Int {
        return children.size
    }

    override fun onBindViewHolder(holder: RecyclerViewAdapter.ViewHolder, position: Int) {
        holder.post.text = children.get(position).data.title
        holder.subreddit.text = "r/" + children.get(position).data.subreddit
        if (children.get(position).data.preview != null) {
            Picasso.get().load(
                children.get(position).data.preview.images.get(0).source.url.replace(
                    "&amp;",
                    "&"
                )
            )
                .resize(600, 600)
                .centerInside()
                .into(holder.image)
        } else {
            holder.image.visibility = View.GONE;
        }
        var link: String = "https://www.reddit.com" + children.get(position).data.permalink
        holder.itemView.setOnClickListener(object : View.OnClickListener {
            override fun onClick(p0: View?) {
                val builder = CustomTabsIntent.Builder()
                builder.setToolbarColor(ContextCompat.getColor(context, R.color.colorPrimary))
                builder.addDefaultShareMenuItem()
                val anotherCustomTab = CustomTabsIntent.Builder().build()
                val requestCode = 100
                val intent = anotherCustomTab.intent
                intent.setData(Uri.parse(link))
                val pendingIntent = PendingIntent.getActivity(
                    context,
                    requestCode,
                    intent,
                    PendingIntent.FLAG_UPDATE_CURRENT
                )
                builder.addMenuItem(children.get(position).data.title, pendingIntent)
                builder.setShowTitle(true)
                builder.setStartAnimations(context, android.R.anim.fade_in, android.R.anim.fade_out)
                builder.setExitAnimations(context, android.R.anim.fade_in, android.R.anim.fade_out)
                val customTabsIntent = builder.build()
                val packageName = customTabHelper.getPackageNameToUse(context, link)
                if (packageName == null) {
                    val intentOpenUri = Intent(context, WebView::class.java)
                    intentOpenUri.putExtra("URL", link)
                    context.startActivity(intentOpenUri)
                } else {
                    customTabsIntent.intent.setPackage(packageName)
                    customTabsIntent.launchUrl(context, Uri.parse(link))
                }
            }

        })
    }

    fun swap(newChildren: MutableList<Children>) {
        children.clear()
        children.addAll(newChildren)
        notifyDataSetChanged()
    }

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        var post = view.post
        var subreddit = view.subbreddit
        var image = view.image
    }
}